runtime/pkg/rduckdb/catalog.go

```diff
@@ -0,0 +1,230 @@
+	t.versionMeta[meta.Version] = meta
+	c.acquireVersion(t, t.currentVersion)
+	if oldVersion != "" {
+		c.releaseVersion(t, oldVersion)
+	}
+
+	c.currentSnapshotID++
+	c.acquireSnapshotUnsafe()
+	if c.currentSnapshotID > 1 {
+		c.releaseSnapshotUnsafe(c.snapshots[c.currentSnapshotID-1])
+	}
+}
+
+// removeTable removes a table from the catalog.
+// If the table is currently used by a snapshot, it will stay in the catalog but marked with deleted=true.
+// When the last snapshot referencing the table is released, the table will be removed completely.
+func (c *catalog) removeTable(name string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	t, ok := c.tables[name]
+	if !ok {
+		c.logger.Debug("table not found in rduckdb catalog", slog.String("name", name))
+	}
+
+	oldVersion := t.currentVersion
+	t.deleted = true
+	t.currentVersion = ""
+
+	c.currentSnapshotID++
+	c.acquireSnapshotUnsafe()
+	c.releaseVersion(t, oldVersion)
+	if c.currentSnapshotID > 1 {
+		c.releaseSnapshotUnsafe(c.snapshots[c.currentSnapshotID-1])
+	}
+}
+
+// listTables returns tableMeta for all active tables present in the catalog.
+func (c *catalog) listTables() []*tableMeta {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	tables := make([]*tableMeta, 0)
+	for _, t := range c.tables {
+		if t.deleted {
+			continue
+		}
+		meta, ok := t.versionMeta[t.currentVersion]
+		if !ok {
+			c.logger.Error("internal error: meta for table not found in catalog", slog.String("name", t.name), slog.String("version", t.currentVersion))
+		}
+		tables = append(tables, meta)
+	}
+	return tables
+}
+
+// acquireSnapshot acquires a snapshot of the current table versions.
+func (c *catalog) acquireSnapshot() *snapshot {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.acquireSnapshotUnsafe()
+}
+
+func (c *catalog) acquireSnapshotUnsafe() *snapshot {
+	s, ok := c.snapshots[c.currentSnapshotID]
+	if ok {
+		s.referenceCount++
+		return s
+	}
+	// first acquire
+	s = &snapshot{
+		id:             c.currentSnapshotID,
+		referenceCount: 1,
+		tables:         make([]*tableMeta, 0),
+	}
+	for _, t := range c.tables {
+		if t.deleted {
+			continue
+		}
+
+		meta, ok := t.versionMeta[t.currentVersion]
+		if !ok {
+			c.logger.Error("internal error: meta for table not found in catalog", slog.String("name", t.name), slog.String("version", t.currentVersion))
+		}
+		s.tables = append(s.tables, meta)
+		c.acquireVersion(t, t.currentVersion)
+	}
+	c.snapshots[c.currentSnapshotID] = s
+	return s
+}
+
+// releaseSnapshot releases a snapshot of table versions.
+func (c *catalog) releaseSnapshot(s *snapshot) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.releaseSnapshotUnsafe(s)
+}
+
+func (c *catalog) releaseSnapshotUnsafe(s *snapshot) {
+	s.referenceCount--
+	if s.referenceCount > 0 {
+		return
+	}
+
+	for _, meta := range s.tables {
+		t, ok := c.tables[meta.Name]
+		if !ok {
+			c.logger.Error("internal error: table not found in catalog", slog.String("name", t.name), slog.String("version", t.currentVersion))
```

Comment:
This does not return but just logs an error so it will anyways panic. I previously saw no alerts being triggered when panic so added an error log.
I can add an explicit panic. May be some misconfiguration in alerts.